mycore-base/src/main/java/org/mycore/services/http/MCRQueryParameter.java

```diff
@@ -0,0 +1,121 @@
+     * Let name have the value of string and let value be the empty string.
+     *
+     * @param encodedString encoded string as stated above
+     * @return a record with name and value as determined by above algorithm
+     */
+    public static MCRQueryParameter ofEncodedString(String encodedString) {
+        String[] parts = encodedString.split("=");
+        switch (parts.length) {
+            case 1:
+                return new MCRQueryParameter(URLDecoder.decode(parts[0], StandardCharsets.UTF_8), "");
+            case 2:
+                return new MCRQueryParameter(URLDecoder.decode(parts[0], StandardCharsets.UTF_8),
+                    URLDecoder.decode(parts[1], StandardCharsets.UTF_8));
+            default:
+                throw new IllegalArgumentException("Could not parse parameter: " + encodedString);
+        }
+    }
+
+    /**
+     * Returnes an encoded String as described in {@link #ofEncodedString(String)}.
+     * <p>
+     * {@snippet :
+     * MCRQueryParameter inst1 = new MCRQueryParameter(name, value);
+     * String encodedString = inst1.toEncodedString();
+     * MCRQueryParameter inst2 = MCRQueryParameter.ofEncodedString(encodedString);
+     * boolean allwaysTrue = inst1.equals(inst2);
+     * }
+     * @return encoded String as definied for <code>application/x-www-form-urlencoded</code>
+     */
+    public String toEncodedString() {
+        if (value().isEmpty()) {
+            return URLEncoder.encode(this.name(), StandardCharsets.UTF_8);
+        }
+        return URLEncoder.encode(this.name(), StandardCharsets.UTF_8) + "="
+            + URLEncoder.encode(this.value(), StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Parses a <code>query</code> string as defined for <code>application/x-www-form-urlencoded</code>
+     *
+     * @param queryString The encoded string, may start with a '?'
+     * @return A list of parameters
+     */
+    public static List<MCRQueryParameter> parse(String queryString) {
+        String[] parts = removeQuestionMark(queryString).split("&");
+        return Stream.of(parts)
+            .map(MCRQueryParameter::ofEncodedString)
+            .collect(Collectors.toUnmodifiableList());
+    }
+
+    private static String removeQuestionMark(String queryString) {
+        return !queryString.isEmpty() && queryString.charAt(0) == '?' ? queryString.substring(1) : queryString;
+    }
+
+    /**
+     * Encodes a list of parameters as a query String.
+     *
+     * @param parameters list of parameters
+     * @return The encoded string, empty if parameters are empty
+     * @throws NullPointerException if parameters are empty or contains null values
+     */
+    public static String toQueryString(List<MCRQueryParameter> parameters) {
+        if (parameters.isEmpty()) {
+            return "";
+        }
+        return parameters.stream()
+            .map(MCRQueryParameter::toEncodedString)
+            .collect(Collectors.joining("&", "?", ""));
```

Comment:
query should not start with '?'

Reply:
fixed in https://github.com/MyCoRe-Org/mycore/pull/2212/commits/669e4f6d271ff9a5718a9c3ae8db413811af7266